### ISSUE TITLE
Rectangle Detection Logic

### ISSUE THREAD

**Issue by mahardhikaprajaonehub** [pymupdf/PyMuPDF#2094]
## Describe the bug (mandatory)
I found incorrect logic in the [rectangle detection](https://github.com/pymupdf/PyMuPDF/blob/1.21/fitz/helper-devices.i#L128-L136).

### Case 1
```
line0 = [[2, 2], [6, 2]]
line1 = [[6, 2], [6, 4]]
line2 = [[6, 4], [10, 4]]
```
![image](https://user-images.githubusercontent.com/26212704/204723709-3e2f3851-417b-4369-a46b-76a87c240198.png)

```
ll = line0[0]
lr = line0[1]
ur = line2[0]
ul = line2[1]
```
![image](https://user-images.githubusercontent.com/26212704/204727435-ecd4e59b-bbad-4180-b8d2-7ad2f5d92e13.png)

Passed the detection test
> ll.y == lr.y
> lr.x == ur.x
> ur.y == ul.y

### Case 2
```
line0 = [[2, 2], [6, 2]]
line1 = [[6, 2], [6, 4]]
line2 = [[6, 4], [4, 4]]
```
![image](https://user-images.githubusercontent.com/26212704/204735122-7fcca57a-dafd-4ccb-be49-a7f163543a5a.png)

```
ll = line0[0]
lr = line0[1]
ur = line2[0]
ul = line2[1]
```
![image](https://user-images.githubusercontent.com/26212704/204735526-17d71888-c788-4bfe-b7fe-fb236a45b4f5.png)

Passed the detection test
> ll.y == lr.y
> lr.x == ur.x
> ur.y == ul.y

## My Solution
I have 2 solutions to resolve this issue

### Solution 1
Only works if the line0 is horizontal
```C
if (ll.y != lr.y || ll.x != ul.x || ur.y != ul.y || ur.x != lr.x) {
	goto drop_out;
}
```

### Solution 2
It can also detect rectangle if the line0 is vertical
```C
if (ll.y == lr.y) {
	if (ll.x != ul.x || ur.y != ul.y || ur.x != lr.x) {
		goto drop_out;
	}
}
else if (ll.x == lr.x) {
	if (ll.y != ul.y || ur.x != ul.x || ur.y != lr.y) {
		goto drop_out;
	}
	// do swap lr => ul
	// It is required to keep orientation detection work properly
	fz_point tmp = lr;
	lr = ul;
	ul = tmp;
}
else {
	goto drop_out;
}
```

## Your configuration (mandatory)
3.9.9 (tags/v3.9.9:ccb0e6a, Nov 15 2021, 18:08:50) [MSC v.1929 64 bit (AMD64)] 
 win32 
 
PyMuPDF 1.21.0: Python bindings for the MuPDF 1.21.0 library.
Version date: 2022-11-08 00:00:01.
Built for Python 3.9 on win32 (64-bit).

**Comment by JorjMcKie** [comment]
Thanks for the report.
You are right: need to also check x-coordinate equality.

BTW: There is no intention to detect arbitrary rectangle-like things. Detection delivers a rectangle if and only if all of the following conditions are met:
1. Three consecutive lines
2. Lines 1 and 3 must be horizontal, line 2 must be vertical
3. The line commands must be followed by a "close path" command

While method `.get_drawings()` works for all supported document types, the reason of these conditions is how the PDF rectangle drawing command `"re"` is decomposed into single line drawings.

Depending on whether line 1 is "above" or "below" line 3, the rectangle **_orientation_** is determined: **_below_** means anti-clockwise, **_above_** is clockwise.

**Comment by julian-smith-artifex-com** [comment]
Fixed in PyMuPDF-1.21.1.
